{
  "comments": [
    {
      "key": {
        "uuid": "9a437d4f_4f774ed5",
        "filename": "src/com/cyanogenmod/updater/service/DownloadCompleteIntentService.java",
        "patchSetId": 1
      },
      "lineNbr": 94,
      "author": {
        "id": 3671
      },
      "writtenOn": "2017-01-28T22:52:20Z",
      "side": 0,
      "message": "How can running this here corrupt the download? If that\u0027s truly what\u0027s happening, it should be corrupting when you run it later too, thereby verifying it and subsequently corrupting it.",
      "range": {
        "startLine": 94,
        "startChar": 16,
        "endLine": 94,
        "endChar": 31
      },
      "revId": "4b122c0c712668a43940039299d422e7cc4f7352",
      "serverId": "1ec6b3db-b2c1-4fa4-84a2-4c7efe89ba71",
      "unresolved": false
    },
    {
      "key": {
        "uuid": "9a437d4f_4f5e2e47",
        "filename": "src/com/cyanogenmod/updater/service/DownloadCompleteIntentService.java",
        "patchSetId": 1
      },
      "lineNbr": 94,
      "author": {
        "id": 12604
      },
      "writtenOn": "2017-01-28T23:19:44Z",
      "side": 0,
      "message": "It seems that the output stream is still writing to destFile when mDm.remove is called and it interrupts the write somehow.  Before the new download manager commit was made, the previous code also did not call mDm.remove right before verifyPackage.\n\nIt is not the verifyPackage that is corrupting the file.  The problem is that the output stream does not finish writing to destFile.",
      "parentUuid": "9a437d4f_4f774ed5",
      "range": {
        "startLine": 94,
        "startChar": 16,
        "endLine": 94,
        "endChar": 31
      },
      "revId": "4b122c0c712668a43940039299d422e7cc4f7352",
      "serverId": "1ec6b3db-b2c1-4fa4-84a2-4c7efe89ba71",
      "unresolved": false
    },
    {
      "key": {
        "uuid": "9a437d4f_1e73a84e",
        "filename": "src/com/cyanogenmod/updater/service/DownloadCompleteIntentService.java",
        "patchSetId": 9
      },
      "lineNbr": 84,
      "author": {
        "id": 7880
      },
      "writtenOn": "2017-02-01T11:41:17Z",
      "side": 0,
      "message": "According to the docs, DownloadManager.openDownloadedFile() is the only method in the original code that can throw an IOException, namely a FileNotFoundException.\n\nIf so, maybe something like this is simpler. I think the null checks aren\u0027t needed for this as well:\n\n ParcelFileDescriptor file \u003d null;\n try {\n     ParcelFileDescriptor file \u003d mDm.openDownloadedFile(id);\n } catch (FileNotFoundException e) {\n     displayErrorResult(updateIntent, R.string.unable_to_download_file);\n     mDm.remove(id);\n     return;\n }\n\n FileInputStream inStream \u003d new FileInputStream(file.getFileDescriptor());\n FileChannel inChannel \u003d inStream.getChannel();\n FileOutputStream outStream \u003d new FileOutputStream(destFile);\n FileChannel outChannel \u003d outStream.getChannel();\n inChannel.transferTo(0, inChannel.size(), outChannel);\n\n try {\n     inChannel.close();\n     outChannel.close();\n     inStream.close();\n     outStream.close();\n } catch (IOException e) {\n     // Ignore\n }\n\n mDm.remove(id);",
      "revId": "4b122c0c712668a43940039299d422e7cc4f7352",
      "serverId": "1ec6b3db-b2c1-4fa4-84a2-4c7efe89ba71",
      "unresolved": false
    }
  ]
}